## Setup

> 📖 Read the full article: [Great Tables: Publication-Ready Tables from Polars and Pandas DataFrames](https://codecut.ai/great-tables-python/)


Great Tables works with both pandas and Polars DataFrames. We'll use Polars in this tutorial:

```bash
pip install great_tables polars selenium
```

Selenium is required for exporting tables as PNG images.

> New to Polars? See our [Polars vs. Pandas comparison](https://codecut.ai/polars-vs-pandas-a-fast-multi-core-alternative-for-dataframes/) for an introduction.

Great Tables includes built-in sample datasets. We'll use the `sp500` dataset containing historical S&P 500 stock data:

In [6]:
from great_tables import GT
from great_tables.data import sp500
import polars as pl

# Preview the raw data
sp500_df = pl.from_pandas(sp500)
print(sp500_df.head(5))

shape: (5, 7)
┌────────────┬───────────┬───────────┬─────────┬───────────┬──────────┬───────────┐
│ date       ┆ open      ┆ high      ┆ low     ┆ close     ┆ volume   ┆ adj_close │
│ ---        ┆ ---       ┆ ---       ┆ ---     ┆ ---       ┆ ---      ┆ ---       │
│ str        ┆ f64       ┆ f64       ┆ f64     ┆ f64       ┆ f64      ┆ f64       │
╞════════════╪═══════════╪═══════════╪═════════╪═══════════╪══════════╪═══════════╡
│ 2015-12-31 ┆ 2060.5901 ┆ 2062.54   ┆ 2043.62 ┆ 2043.9399 ┆ 2.6553e9 ┆ 2043.9399 │
│ 2015-12-30 ┆ 2077.3401 ┆ 2077.3401 ┆ 2061.97 ┆ 2063.3601 ┆ 2.3674e9 ┆ 2063.3601 │
│ 2015-12-29 ┆ 2060.54   ┆ 2081.5601 ┆ 2060.54 ┆ 2078.3601 ┆ 2.5420e9 ┆ 2078.3601 │
│ 2015-12-28 ┆ 2057.77   ┆ 2057.77   ┆ 2044.2  ┆ 2056.5    ┆ 2.4925e9 ┆ 2056.5    │
│ 2015-12-24 ┆ 2063.52   ┆ 2067.3601 ┆ 2058.73 ┆ 2060.99   ┆ 1.4119e9 ┆ 2060.99   │
└────────────┴───────────┴───────────┴─────────┴───────────┴──────────┴───────────┘


```text
shape: (5, 7)
┌────────────┬───────────┬───────────┬─────────┬───────────┬──────────┬───────────┐
│ date       ┆ open      ┆ high      ┆ low     ┆ close     ┆ volume   ┆ adj_close │
│ ---        ┆ ---       ┆ ---       ┆ ---     ┆ ---       ┆ ---      ┆ ---       │
│ str        ┆ f64       ┆ f64       ┆ f64     ┆ f64       ┆ f64      ┆ f64       │
╞════════════╪═══════════╪═══════════╪═════════╪═══════════╪══════════╪═══════════╡
│ 2015-12-31 ┆ 2060.5901 ┆ 2062.54   ┆ 2043.62 ┆ 2043.9399 ┆ 2.6553e9 ┆ 2043.9399 │
│ 2015-12-30 ┆ 2077.3401 ┆ 2077.3401 ┆ 2061.97 ┆ 2063.3601 ┆ 2.3674e9 ┆ 2063.3601 │
│ 2015-12-29 ┆ 2060.54   ┆ 2081.5601 ┆ 2060.54 ┆ 2078.3601 ┆ 2.5420e9 ┆ 2078.3601 │
│ 2015-12-28 ┆ 2057.77   ┆ 2057.77   ┆ 2044.2  ┆ 2056.5    ┆ 2.4925e9 ┆ 2056.5    │
│ 2015-12-24 ┆ 2063.52   ┆ 2067.3601 ┆ 2058.73 ┆ 2060.99   ┆ 1.4119e9 ┆ 2060.99   │
└────────────┴───────────┴───────────┴─────────┴───────────┴──────────┴───────────┘
```

The raw output shows:

- Unformatted decimals (e.g., 2060.5901)
- Large integers without separators (e.g., 2.6553e9)
- Dates as plain strings (e.g., "2015-12-31")

Let's transform this into a readable table.

## Value Formatting

Great Tables provides `fmt_*` methods to format values. Here's how to format currencies, numbers, and dates:

In [7]:
from great_tables import GT
from great_tables.data import sp500

# Filter to a specific date range
start_date = "2010-06-07"
end_date = "2010-06-14"
sp500_mini = sp500[(sp500["date"] >= start_date) & (sp500["date"] <= end_date)]

stock_price_table = (
    GT(sp500_mini)
    .fmt_currency(columns=["open", "high", "low", "close"])
    .fmt_date(columns="date", date_style="wd_m_day_year")
    .fmt_number(columns="volume", compact=True)
    .cols_hide(columns="adj_close")
)
stock_price_table

GT(_tbl_data=            date       open     high      low      close        volume  \
1398  2010-06-14  1095.0000  1105.91  1089.03  1089.6300  4.425830e+09   
1399  2010-06-11  1082.6500  1092.25  1077.12  1091.6000  4.059280e+09   
1400  2010-06-10  1058.7700  1087.85  1058.77  1086.8400  5.144780e+09   
1401  2010-06-09  1062.7500  1077.74  1052.25  1055.6899  5.983200e+09   
1402  2010-06-08  1050.8101  1063.15  1042.17  1062.0000  6.192750e+09   
1403  2010-06-07  1065.8400  1071.36  1049.86  1050.4700  5.467560e+09   

      adj_close  
1398  1089.6300  
1399  1091.6000  
1400  1086.8400  
1401  1055.6899  
1402  1062.0000  
1403  1050.4700  , _body=<great_tables._gt_data.Body object at 0x11ca61890>, _boxhead=Boxhead([ColInfo(var='date', type=<ColInfoTypeEnum.default: 1>, column_label='date', column_align='right', column_width=None), ColInfo(var='open', type=<ColInfoTypeEnum.default: 1>, column_label='open', column_align='right', column_width=None), ColInfo(var='high', type=<ColInfoTypeEnum.default: 1>, column_label='high', column_align='right', column_width=None), ColInfo(var='low', type=<ColInfoTypeEnum.default: 1>, column_label='low', column_align='right', column_width=None), ColInfo(var='close', type=<ColInfoTypeEnum.default: 1>, column_label='close', column_align='right', column_width=None), ColInfo(var='volume', type=<ColInfoTypeEnum.default: 1>, column_label='volume', column_align='right', column_width=None), ColInfo(var='adj_close', type=<ColInfoTypeEnum.hidden: 4>, column_label='adj_close', column_align='right', column_width=None)]), _stub=<great_tables._gt_data.Stub object at 0x11daa0790>, _spanners=Spanners([]), _heading=Heading(title=None, subtitle=None, preheader=None), _stubhead=None, _summary_rows=<great_tables._gt_data.SummaryRows object at 0x11ca5f990>, _summary_rows_grand=<great_tables._gt_data.SummaryRows object at 0x11da7b690>, _source_notes=[], _footnotes=[], _styles=[], _locale=<great_tables._gt_data.Locale object at 0x11da78410>, _formats=[<great_tables._gt_data.FormatInfo object at 0x11dbb3d90>, <great_tables._gt_data.FormatInfo object at 0x11dbb3690>, <great_tables._gt_data.FormatInfo object at 0x11dbb3650>], _substitutions=[], _options=Options(table_id=OptionsInfo(scss=False, category='table', type='value', value=None), table_caption=OptionsInfo(scss=False, category='table', type='value', value=None), table_width=OptionsInfo(scss=True, category='table', type='px', value='auto'), table_layout=OptionsInfo(scss=True, category='table', type='value', value='fixed'), table_margin_left=OptionsInfo(scss=True, category='table', type='px', value='auto'), table_margin_right=OptionsInfo(scss=True, category='table', type='px', value='auto'), table_background_color=OptionsInfo(scss=True, category='table', type='value', value='#FFFFFF'), table_additional_css=OptionsInfo(scss=False, category='table', type='values', value=[]), table_font_names=OptionsInfo(scss=False, category='table', type='values', value=['-apple-system', 'BlinkMacSystemFont', 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Helvetica Neue', 'Fira Sans', 'Droid Sans', 'Arial', 'sans-serif']), table_font_size=OptionsInfo(scss=True, category='table', type='px', value='16px'), table_font_weight=OptionsInfo(scss=True, category='table', type='value', value='normal'), table_font_style=OptionsInfo(scss=True, category='table', type='value', value='normal'), table_font_color=OptionsInfo(scss=True, category='table', type='value', value='#333333'), table_font_color_light=OptionsInfo(scss=True, category='table', type='value', value='#FFFFFF'), table_border_top_include=OptionsInfo(scss=False, category='table', type='boolean', value=True), table_border_top_style=OptionsInfo(scss=True, category='table', type='value', value='solid'), table_border_top_width=OptionsInfo(scss=True, category='table', type='px', value='2px'), table_border_top_color=OptionsInfo(scss=True, category='table', type='value', value='#A8A8A8'), table_border_right_style=OptionsInfo(

![Stock Price Table](https://codecut.ai/wp-content/uploads/2025/12/stock_price_table.png)

In this example:

- `fmt_currency()` adds dollar signs and formats decimals (e.g., $1,065.84)
- `fmt_date()` converts date strings to readable format (e.g., "Mon, Jun 7, 2010")
- `fmt_number()` with `compact=True` converts large numbers to compact format (e.g., 5.47B)
- `cols_hide()` removes the redundant `adj_close` column

To export the table for reports, use the `save()` method:

In [8]:
stock_price_table.save("stock_price_table.png")  # Supports .png, .bmp, .pdf

GT(_tbl_data=            date       open     high      low      close        volume  \
1398  2010-06-14  1095.0000  1105.91  1089.03  1089.6300  4.425830e+09   
1399  2010-06-11  1082.6500  1092.25  1077.12  1091.6000  4.059280e+09   
1400  2010-06-10  1058.7700  1087.85  1058.77  1086.8400  5.144780e+09   
1401  2010-06-09  1062.7500  1077.74  1052.25  1055.6899  5.983200e+09   
1402  2010-06-08  1050.8101  1063.15  1042.17  1062.0000  6.192750e+09   
1403  2010-06-07  1065.8400  1071.36  1049.86  1050.4700  5.467560e+09   

      adj_close  
1398  1089.6300  
1399  1091.6000  
1400  1086.8400  
1401  1055.6899  
1402  1062.0000  
1403  1050.4700  , _body=<great_tables._gt_data.Body object at 0x11ca61890>, _boxhead=Boxhead([ColInfo(var='date', type=<ColInfoTypeEnum.default: 1>, column_label='date', column_align='right', column_width=None), ColInfo(var='open', type=<ColInfoTypeEnum.default: 1>, column_label='open', column_align='right', column_width=None), ColInfo(var='high', type=<ColInfoTypeEnum.default: 1>, column_label='high', column_align='right', column_width=None), ColInfo(var='low', type=<ColInfoTypeEnum.default: 1>, column_label='low', column_align='right', column_width=None), ColInfo(var='close', type=<ColInfoTypeEnum.default: 1>, column_label='close', column_align='right', column_width=None), ColInfo(var='volume', type=<ColInfoTypeEnum.default: 1>, column_label='volume', column_align='right', column_width=None), ColInfo(var='adj_close', type=<ColInfoTypeEnum.hidden: 4>, column_label='adj_close', column_align='right', column_width=None)]), _stub=<great_tables._gt_data.Stub object at 0x11daa0790>, _spanners=Spanners([]), _heading=Heading(title=None, subtitle=None, preheader=None), _stubhead=None, _summary_rows=<great_tables._gt_data.SummaryRows object at 0x11ca5f990>, _summary_rows_grand=<great_tables._gt_data.SummaryRows object at 0x11da7b690>, _source_notes=[], _footnotes=[], _styles=[], _locale=<great_tables._gt_data.Locale object at 0x11da78410>, _formats=[<great_tables._gt_data.FormatInfo object at 0x11dbb3d90>, <great_tables._gt_data.FormatInfo object at 0x11dbb3690>, <great_tables._gt_data.FormatInfo object at 0x11dbb3650>], _substitutions=[], _options=Options(table_id=OptionsInfo(scss=False, category='table', type='value', value=None), table_caption=OptionsInfo(scss=False, category='table', type='value', value=None), table_width=OptionsInfo(scss=True, category='table', type='px', value='auto'), table_layout=OptionsInfo(scss=True, category='table', type='value', value='fixed'), table_margin_left=OptionsInfo(scss=True, category='table', type='px', value='auto'), table_margin_right=OptionsInfo(scss=True, category='table', type='px', value='auto'), table_background_color=OptionsInfo(scss=True, category='table', type='value', value='#FFFFFF'), table_additional_css=OptionsInfo(scss=False, category='table', type='values', value=[]), table_font_names=OptionsInfo(scss=False, category='table', type='values', value=['-apple-system', 'BlinkMacSystemFont', 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Helvetica Neue', 'Fira Sans', 'Droid Sans', 'Arial', 'sans-serif']), table_font_size=OptionsInfo(scss=True, category='table', type='px', value='16px'), table_font_weight=OptionsInfo(scss=True, category='table', type='value', value='normal'), table_font_style=OptionsInfo(scss=True, category='table', type='value', value='normal'), table_font_color=OptionsInfo(scss=True, category='table', type='value', value='#333333'), table_font_color_light=OptionsInfo(scss=True, category='table', type='value', value='#FFFFFF'), table_border_top_include=OptionsInfo(scss=False, category='table', type='boolean', value=True), table_border_top_style=OptionsInfo(scss=True, category='table', type='value', value='solid'), table_border_top_width=OptionsInfo(scss=True, category='table', type='px', value='2px'), table_border_top_color=OptionsInfo(scss=True, category='table', type='value', value='#A8A8A8'), table_border_right_style=OptionsInfo(

### Formatting Percentages

Use `fmt_percent()` to display decimal values as percentages. Here's some sample data with decimal values:

In [ ]:
import polars as pl
from great_tables import GT

performance_data = pl.DataFrame({
    "metric": ["Revenue Growth", "Profit Margin", "Market Share"],
    "q1": [0.12, 0.08, 0.23],
    "q2": [0.15, 0.09, 0.25],
    "q3": [0.11, 0.07, 0.24]
})
performance_data

The raw decimals are hard to read at a glance. Let's format them as percentages:

In [ ]:
percent_table = (
    GT(performance_data, rowname_col="metric")
    .fmt_percent(columns=["q1", "q2", "q3"], decimals=1)
)
percent_table

![Percent Table](https://codecut.ai/wp-content/uploads/2025/12/percent_table.png)

The percentages are now much more readable! Values like 0.12 become "12.0%" automatically.

## Table Structure

Professional tables need clear headers, grouped columns, and source attribution. Great Tables provides methods for each structural component.

### Adding Headers and Source Notes

Use `tab_header()` for titles and `tab_source_note()` for attribution. Let's start with our S&P 500 data:

In [ ]:
from great_tables import GT, md
from great_tables.data import sp500
import polars as pl

sp500_pl = pl.from_pandas(sp500)
sp500_mini = sp500_pl.filter(
    (pl.col("date") >= "2010-06-07") & (pl.col("date") <= "2010-06-14")
)
print(sp500_mini)

shape: (6, 7)
┌────────────┬───────────┬─────────┬─────────┬───────────┬──────────┬───────────┐
│ date       ┆ open      ┆ high    ┆ low     ┆ close     ┆ volume   ┆ adj_close │
│ ---        ┆ ---       ┆ ---     ┆ ---     ┆ ---       ┆ ---      ┆ ---       │
│ str        ┆ f64       ┆ f64     ┆ f64     ┆ f64       ┆ f64      ┆ f64       │
╞════════════╪═══════════╪═════════╪═════════╪═══════════╪══════════╪═══════════╡
│ 2010-06-14 ┆ 1095.0    ┆ 1105.91 ┆ 1089.03 ┆ 1089.63   ┆ 4.4258e9 ┆ 1089.63   │
│ 2010-06-11 ┆ 1082.65   ┆ 1092.25 ┆ 1077.12 ┆ 1091.6    ┆ 4.0593e9 ┆ 1091.6    │
│ 2010-06-10 ┆ 1058.77   ┆ 1087.85 ┆ 1058.77 ┆ 1086.84   ┆ 5.1448e9 ┆ 1086.84   │
│ 2010-06-09 ┆ 1062.75   ┆ 1077.74 ┆ 1052.25 ┆ 1055.6899 ┆ 5.9832e9 ┆ 1055.6899 │
│ 2010-06-08 ┆ 1050.8101 ┆ 1063.15 ┆ 1042.17 ┆ 1062.0    ┆ 6.1928e9 ┆ 1062.0    │
│ 2010-06-07 ┆ 1065.84   ┆ 1071.36 ┆ 1049.86 ┆ 1050.47   ┆ 5.4676e9 ┆ 1050.47   │
└────────────┴───────────┴─────────┴─────────┴───────────┴──────────┴───────────┘

The table lacks context about what the data represents. Let's add a title and source:

In [ ]:
header_table = (
    GT(sp500_mini)
    .tab_header(
        title="S&P 500 Daily Performance",
        subtitle="June 7-14, 2010"
    )
    .fmt_currency(columns=["open", "high", "low", "close"])
    .fmt_date(columns="date", date_style="wd_m_day_year")
    .fmt_number(columns="volume", compact=True)
    .cols_hide(columns="adj_close")
    .tab_source_note(source_note=md("**Source**: Historical market data"))
)
header_table

![Header Table](https://codecut.ai/wp-content/uploads/2025/12/header_table.png)

In this example:

- `tab_header()` adds "S&P 500 Daily Performance" as the title and "June 7-14, 2010" as the subtitle
- `tab_source_note()` adds "**Source**: Historical market data" at the bottom
- `md()` enables markdown formatting for bold text

### Grouping Columns with Spanners

Column spanners group related columns under a shared label. Here's some quarterly sales data:

In [ ]:
import polars as pl
from great_tables import GT

sales_data = pl.DataFrame({
    "product": ["Laptop", "Phone", "Tablet"],
    "q1_rev": [125000, 89000, 45000],
    "q2_rev": [132000, 95000, 48000],
    "q1_units": [450, 1200, 380],
    "q2_units": [475, 1350, 410]
})
print(sales_data)

┌─────────┬────────┬────────┬──────────┬──────────┐
│ product ┆ q1_rev ┆ q2_rev ┆ q1_units ┆ q2_units │
│ ---     ┆ ---    ┆ ---    ┆ ---      ┆ ---      │
│ str     ┆ i64    ┆ i64    ┆ i64      ┆ i64      │
╞═════════╪════════╪════════╪══════════╪══════════╡
│ Laptop  ┆ 125000 ┆ 132000 ┆ 450      ┆ 475      │
│ Phone   ┆ 89000  ┆ 95000  ┆ 1200     ┆ 1350     │
│ Tablet  ┆ 45000  ┆ 48000  ┆ 380      ┆ 410      │
└─────────┴────────┴────────┴──────────┴──────────┘

The column names like `q1_rev` and `q1_units` don't clearly show their relationship. Let's group them with spanners:

In [ ]:
spanner_table = (
    GT(sales_data, rowname_col="product")
    .tab_header(title="Quarterly Sales Report")
    .tab_spanner(label="Revenue ($)", columns=["q1_rev", "q2_rev"])
    .tab_spanner(label="Units Sold", columns=["q1_units", "q2_units"])
    .fmt_currency(columns=["q1_rev", "q2_rev"], decimals=0)
    .fmt_number(columns=["q1_units", "q2_units"], use_seps=True)
    .cols_label(
        q1_rev="Q1",
        q2_rev="Q2",
        q1_units="Q1",
        q2_units="Q2"
    )
    .tab_stubhead(label="Product")
)
spanner_table

![Spanner Table](https://codecut.ai/wp-content/uploads/2025/12/spanner_table.png)

In this example:

- `tab_spanner()` creates "Revenue ($)" and "Units Sold" headers that span multiple columns
- `cols_label()` renames columns like `q1_rev` to "Q1"
- `tab_stubhead()` labels the row name column as "Product"

## Data Coloring

The `data_color()` method applies color scales to cells based on their values, creating heatmap-style visualizations. Here's some regional performance data:

In [ ]:
import polars as pl
from great_tables import GT

performance = pl.DataFrame({
    "region": ["North", "South", "East", "West"],
    "revenue": [125000, 98000, 145000, 112000],
    "growth": [0.15, -0.05, 0.22, 0.08]
})
print(performance)

┌────────┬─────────┬────────┐
│ region ┆ revenue ┆ growth │
│ ---    ┆ ---     ┆ ---    │
│ str    ┆ i64     ┆ f64    │
╞════════╪═════════╪════════╡
│ North  ┆ 125000  ┆ 0.15   │
│ South  ┆ 98000   ┆ -0.05  │
│ East   ┆ 145000  ┆ 0.22   │
│ West   ┆ 112000  ┆ 0.08   │
└────────┴─────────┴────────┘

The raw numbers make it hard to spot which regions are performing well. Let's add color scales:

In [ ]:
color_table = (
    GT(performance, rowname_col="region")
    .fmt_currency(columns="revenue", decimals=0)
    .fmt_percent(columns="growth", decimals=1)
    .data_color(
        columns="revenue",
        palette="Blues"
    )
    .data_color(
        columns="growth",
        palette=["red", "white", "green"],
        domain=[-0.1, 0.25]
    )
)
color_table

![Color Table](https://codecut.ai/wp-content/uploads/2025/12/color_table.png)

Now high performers stand out immediately! In this example:

- `palette="Blues"` applies a blue gradient to revenue (darker = higher values like $145,000)
- `palette=["red", "white", "green"]` creates a diverging scale for growth (red for -5.0%, green for 22.0%)
- `domain=[-0.1, 0.25]` sets the min/max range for the color scale

## Nanoplots

Nanoplots embed small visualizations directly in table cells. They're useful for showing trends without creating separate charts.

### Creating Line Nanoplots

To use nanoplots, your data column must contain space-separated numeric values:

In [ ]:
import polars as pl
from great_tables import GT

# Create data with trend values as space-separated strings
kpi_data = pl.DataFrame({
    "metric": ["Revenue", "Users", "Conversion Rate"],
    "current": [125000.0, 45000.0, 3.2],
    "trend": [
        "95 102 98 115 125",
        "38 40 42 43 45",
        "2.8 2.9 3.0 3.1 3.2"
    ]
})

kpi_table = (
    GT(kpi_data, rowname_col="metric")
    .fmt_nanoplot(columns="trend", plot_type="line")
    .fmt_number(columns="current", compact=True)
    .tab_header(title="Weekly KPI Dashboard")
)
kpi_table

![KPI Table](https://codecut.ai/wp-content/uploads/2025/12/kpi_table.png)

The sparklines make trends instantly visible! `fmt_nanoplot()` transforms space-separated values like "95 102 98 115 125" into inline charts.

Hover over the chart to see individual data points.

### Adding Reference Lines

Reference lines provide context by showing averages, medians, or custom thresholds:

In [ ]:
import polars as pl
from great_tables import GT

trend_data = pl.DataFrame({
    "stock": ["AAPL", "GOOGL", "MSFT"],
    "prices": [
        "150 155 148 160 165 158 170",
        "120 118 122 125 128 130 127",
        "280 285 275 290 295 288 300"
    ]
})

stock_trend_table = (
    GT(trend_data, rowname_col="stock")
    .fmt_nanoplot(
        columns="prices",
        plot_type="line",
        reference_line="mean"
    )
    .tab_header(title="Weekly Stock Prices")
)
stock_trend_table

![Stock Trend Table](https://codecut.ai/wp-content/uploads/2025/12/stock_trend_table.png)

The `reference_line="mean"` parameter adds a horizontal line at the average value. Other options include `"median"`, `"min"`, `"max"`, `"q1"`, and `"q3"`.

### Bar Nanoplots

Use `plot_type="bar"` for comparing discrete values:

In [ ]:
import polars as pl
from great_tables import GT

monthly_data = pl.DataFrame({
    "category": ["Electronics", "Clothing", "Food"],
    "sales": [
        "45 52 48 55 60 58",
        "30 28 35 32 38 40",
        "20 22 21 25 24 26"
    ]
})

bar_chart_table = (
    GT(monthly_data, rowname_col="category")
    .fmt_nanoplot(columns="sales", plot_type="bar")
    .tab_header(title="Monthly Sales by Category")
)
bar_chart_table

![Bar Chart Table](https://codecut.ai/wp-content/uploads/2025/12/bar_chart_table.png)

### Customizing Nanoplot Appearance

Pass styling options via `nanoplot_options()`:

- **Line**: `data_line_stroke_color` (e.g., "steelblue")
- **Points**: `data_point_fill_color`, `data_point_stroke_color`
- **Area**: `data_area_fill_color` (e.g., "lightblue")

In [ ]:
from great_tables import GT, nanoplot_options
import polars as pl

trend_data = pl.DataFrame({
    "metric": ["Growth", "Engagement"],
    "values": ["10 15 12 18 22 20", "5 8 6 9 11 10"]
})

styled_nanoplot_table = (
    GT(trend_data, rowname_col="metric")
    .fmt_nanoplot(
        columns="values",
        plot_type="line",
        reference_line="mean",
        options=nanoplot_options(
            data_line_stroke_color="steelblue",
            data_point_fill_color="white",
            data_point_stroke_color="steelblue",
            data_area_fill_color="lightblue"
        )
    )
)
styled_nanoplot_table

![Styled Nanoplot Table](https://codecut.ai/wp-content/uploads/2025/12/styled_nanoplot_table.png)

## Conditional Styling

The `tab_style()` method applies formatting to cells based on conditions. Combined with Polars expressions, you can create data-driven styling rules.

### Basic Conditional Styling

Here's some product sales data with mixed growth values:

In [ ]:
from great_tables import GT, style, loc
import polars as pl

sales = pl.DataFrame({
    "product": ["Laptop", "Phone", "Tablet", "Monitor"],
    "revenue": [125000, 89000, 45000, 32000],
    "growth": [0.15, -0.05, 0.22, -0.08]
})
print(sales)

┌─────────┬──────────┬────────┐
│ product ┆ revenue ┆ growth │
│ ---     ┆ ---     ┆ ---    │
│ str     ┆ i64     ┆ f64    │
╞═════════╪═════════╪════════╡
│ Laptop  ┆ 125000  ┆ 0.15   │
│ Phone   ┆ 89000   ┆ -0.05  │
│ Tablet  ┆ 45000   ┆ 0.22   │
│ Monitor ┆ 32000   ┆ -0.08  │
└─────────┴─────────┴────────┘

Some products have positive growth, others negative. Let's use `tab_style()` with Polars expressions to apply conditional colors:

In [ ]:
conditional_table = (
    GT(sales, rowname_col="product")
    .fmt_currency(columns="revenue", decimals=0)
    .fmt_percent(columns="growth", decimals=1)
    .tab_style(
        style=[
            style.fill(color="lightgreen"),
            style.text(weight="bold")
        ],
        locations=loc.body(
            columns="growth",
            rows=pl.col("growth") > 0
        )
    )
    .tab_style(
        style=[
            style.fill(color="lightcoral"),
            style.text(weight="bold")
        ],
        locations=loc.body(
            columns="growth",
            rows=pl.col("growth") < 0
        )
    )
)
conditional_table

![Conditional Table](https://codecut.ai/wp-content/uploads/2025/12/conditional_table.png)

The styling makes values immediately visible:

- `pl.col("growth") > 0` - selects rows with positive growth
- `pl.col("growth") < 0` - selects rows with negative growth